### PR TITLE
Option to execute multiple runs non-interleaved

### PR DESCRIPTION
**Pull request by bauersimon**

### CODE REVIEW

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 1]
```diff
@@ -48,6 +48,8 @@ type Evaluate struct {
 	ResultPath string `long:"result-path" description:"Directory path where results should be written to. The placeholder \"%datetime%\" can be used for the current date and time." default:"evaluation-%datetime%"`
 	// Runs holds the number of runs to perform.
 	Runs uint `long:"runs" description:"Number of runs to perform." default:"1"`
+	// RunsInterleavedDisabled indicates that interleaved runs are disabled.
+	RunsInterleavedDisabled bool `long:"disable-interleaved" description:"Multiple runs are performed in an interleaved fashion to avoid caching of model responses. Disabling this behavior queries the same model repeatedly."`
```
[the remark this answers is not in the document]
Comment (Munsio) [reply cmd/eval-dev-quality/cmd/evaluate.go:52]: By setting the default value either to true/false the behavior explanation is still the same for the user.

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 2]
```diff
@@ -58,6 +60,22 @@ type Evaluate struct {
...
 
+func (command *Evaluate) runsAtLanguageLevel() uint {
+	if command.RunsInterleavedDisabled {
+		return 1
+	}
+
+	return command.Runs
+}
+
+func (command *Evaluate) runsAtModelLevel() uint {
```
Comment (Munsio) [inline cmd/eval-dev-quality/cmd/evaluate.go:71]: Same for documentation.

File: cmd/eval-dev-quality/cmd/evaluate.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Munsio) [reply cmd/eval-dev-quality/cmd/evaluate.go:52]: I know you want to have `--disable-interleave` as flag but imho it is also better for automation to just have `runs-interleaved=true/false`.

**Review by Munsio — DISMISSED** [review]

File: cmd/eval-dev-quality/cmd/evaluate.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (bauersimon) [reply cmd/eval-dev-quality/cmd/evaluate.go:52]: I cannot set the default value to true: "Could not add arguments group: boolean flag `runs-interleaved' may not have default values, they always default to `false' and can only be turned on" so I guess to work around that we need to make it a string and do the parsing ourselves? :see_no_evil:

[the remark this answers is not in the document]
Comment (Munsio) [reply cmd/eval-dev-quality/cmd/evaluate.go:52]: No the only thing that I could think of would be `run-model-sequential` or something like that. So using the opposite of `interleaved`

[the remark this answers is not in the document]
Comment (bauersimon) [reply cmd/eval-dev-quality/cmd/evaluate.go:52]: I like that one.

**Comment by bauersimon** [comment]
Oh nooo. @Munsio is off today. But he requested changes which I did. But I cannot un-assign him. So this is basically un-mergeable until he is back :face_exhaling:  WHYYY? @zimmski

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 3]
```diff
@@ -48,6 +48,8 @@ type Evaluate struct {
 	ResultPath string `long:"result-path" description:"Directory path where results should be written to. The placeholder \"%datetime%\" can be used for the current date and time." default:"evaluation-%datetime%"`
 	// Runs holds the number of runs to perform.
 	Runs uint `long:"runs" description:"Number of runs to perform." default:"1"`
+	// RunsSequential indicates that interleaved runs are disabled and runs are performed sequentially.
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate.go:51]: Taking this PR.

Comment (zimmski) [reply cmd/eval-dev-quality/cmd/evaluate.go:51]: Nope. too much other stuff.

File: cmd/eval-dev-quality/cmd/evaluate_test.go [hunk 4]
```diff
@@ -576,6 +576,60 @@ func TestEvaluateExecute(t *testing.T) {
...
+				"golang-summed.csv": nil,
+				"models-summed.csv": nil,
+				"README.md":         nil,
+				filepath.Join("symflower_symbolic-execution", "golang", "golang", "plain.log"): func(t *testing.T, filePath, data string) {
+					assert.Equal(t, 3, strings.Count(data, `Evaluating model "symflower/symbolic-execution"`))
+				},
+			},
+		})
+		validate(t, &testCase{
+			Name: "Sequential",
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate_test.go:591]: Test, that this is sequentiel.

Comment (Munsio) [reply cmd/eval-dev-quality/cmd/evaluate_test.go]: The log output validation is different between `interleaved` and `sequential` so we are testing already that the execution is truthfully `sequential`

File: cmd/eval-dev-quality/cmd/evaluate.go (hunk 3 above)
Comment (Munsio) [reply cmd/eval-dev-quality/cmd/evaluate.go:60]: Done

**Comment by ruiAzevedo19** [comment]
@Munsio Review please

File: evaluate/evaluate.go [hunk 5]
```diff
@@ -75,16 +91,26 @@ func Evaluate(ctx *Context) (assessments report.AssessmentPerModelPerLanguagePer
...
-					}
-					if len(ps) > 0 {
-						ctx.Log.Printf("Model %q was not able to solve the %q repository for language %q: %+v", modelID, repositoryPath, languageID, ps)
-						problemsPerModel[modelID] = append(problemsPerModel[modelID], ps...)
-					} else {
-						modelSucceededBasicChecksOfLanguage[model][language] = true
+					for rm := uint(0); rm < ctx.runsAtModelLevel(); rm++ {
+						if err := ResetTemporaryRepository(ctx.Log, temporaryRepositoryPath); err != nil {
+							ctx.Log.Panicf("ERROR: unable to reset temporary repository path: %s", err)
+						}
```
[the remark this answers is not in the document]
Comment (Munsio) [reply evaluate/evaluate.go:97]: That is okay, we need to reset the repository before each model run.

File: evaluate/evaluate.go [hunk 6]
```diff
@@ -46,9 +66,9 @@ func Evaluate(ctx *Context) (assessments report.AssessmentPerModelPerLanguagePer
 	assessments = report.NewAssessmentPerModelPerLanguagePerRepository(ctx.Models, ctx.Languages, ctx.RepositoryPaths)
 	problemsPerModel := map[string][]error{}
 	{
-		for r := uint(0); r < ctx.Runs; r++ {
-			if ctx.Runs > 1 {
-				ctx.Log.Printf("Run %d/%d", r+1, ctx.Runs)
+		for rl := uint(0); rl < ctx.runsAtLanguageLevel(); rl++ {
```
Comment (Munsio) [inline evaluate/evaluate.go:69]: I guess I found a bug with the `TemporaryRepository`. We want to copy the repository for each language only 1 time but currently we are doing it on every run. So 3 runs = 3 x language copies.

We need to iterate over the languages before this loop and create the temporaryRepositoryPaths. We then need to store them inside a map and use the `repositoryPath` as index so we can fetch the correct temporary one during the model run.
The cleanup functions also need to be taken care of in a slice/map and run after everything is done.

Comment (Munsio) [reply evaluate/evaluate.go:69]: This is also true for the loop below where we are not using the `RepositoryPlainName`

**Review by Munsio — CHANGES_REQUESTED** [review]

File: evaluate/evaluate.go [hunk 7]
```diff
@@ -45,28 +65,39 @@ func Evaluate(ctx *Context) (assessments report.AssessmentPerModelPerLanguagePer
...
 	{
-		for r := uint(0); r < ctx.Runs; r++ {
-			if ctx.Runs > 1 {
-				ctx.Log.Printf("Run %d/%d", r+1, ctx.Runs)
+		// Create temporary repositories for each language so the repository is copied only once per language.
+		temporaryRepositories := map[string]*temporaryRepository{}
+		for _, language := range ctx.Languages {
+			repositoryPath := filepath.Join(language.ID(), RepositoryPlainName)
+			temporaryRepositoryPath, cleanup, err := TemporaryRepository(ctx.Log, filepath.Join(ctx.TestdataPath, repositoryPath))
+			if err != nil {
```
Comment (Munsio) [inline evaluate/evaluate.go:80]: I guess I got a little over board with the `cleanup` - you could just `defer` the cleanup after the error check and it should be alright.

Then you do not have to introduce a new type but rather a map[string]string

**Review by Munsio — CHANGES_REQUESTED** [review]

File: evaluate/evaluate.go [hunk 8]
```diff
@@ -75,20 +100,30 @@ func Evaluate(ctx *Context) (assessments report.AssessmentPerModelPerLanguagePer
...
-					if err != nil {
-						ps = append(ps, err)
-					}
-					if len(ps) > 0 {
-						ctx.Log.Printf("Model %q was not able to solve the %q repository for language %q: %+v", modelID, repositoryPath, languageID, ps)
-						problemsPerModel[modelID] = append(problemsPerModel[modelID], ps...)
-					} else {
-						modelSucceededBasicChecksOfLanguage[model][language] = true
+					for rm := uint(0); rm < ctx.runsAtModelLevel(); rm++ {
+						temporaryRepositoryPath := temporaryRepositories[repositoryPath]
```
Comment (Munsio) [inline evaluate/evaluate.go:104]: move that up to line 90 to avoid unnecessary reassignments.

File: evaluate/evaluate.go [hunk 9]
```diff
@@ -126,33 +181,34 @@ func Evaluate(ctx *Context) (assessments report.AssessmentPerModelPerLanguagePer
...
 						continue
 					}
 
-					assessment, ps, err := Repository(ctx.Log, ctx.ResultPath, model, language, temporaryRepositoryPath, repositoryPath)
-					assessments[model][language][repositoryPath].Add(assessment)
-					problemsPerModel[modelID] = append(problemsPerModel[modelID], ps...)
-					if err != nil {
-						ctx.Log.Printf("ERROR: Model %q encountered a hard error for language %q, repository %q: %+v", modelID, languageID, repositoryPath, err)
+					for rm := uint(0); rm < ctx.runsAtModelLevel(); rm++ {
+						temporaryRepositoryPath := temporaryRepositories[repositoryPath]
```
Comment (Munsio) [inline evaluate/evaluate.go:194]: Here too move the assignment up to line 173

**Review by Munsio — CHANGES_REQUESTED** [review]

**Review by Munsio — APPROVED** [review]

**Pull request merged**